teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,431 @@
+     - name: package_version
+       description: version of the uploaded package
+   ```
+   
+   Note some interesting things are happening above already, for example sdist_shaw contains both the filename and the
+   sha - an attempt to group related information.
+   
+   With object results, we could define the above with a bit more structure like this:
+   
+   ```yaml
+     results:
+       - name: sdist
+         description: |
+           The source distribution
+            * sha: The sha256 of the contents
+            * path: Path to the resulting tar.gz
+         schema:
+           type: object
+           properties:
+             sha:
+             path: {}
+       - name: bdist
+         description: |
+           The built distribution
+            * sha: The sha256 of the contents
+            * path: Path to the resulting .whl file
+         schema:
+           type: object
+           properties:
+             sha: {}
+             path: {}
+       - name: package
+         description: |
+           Details about the created package
+            * name: The name of the package
+            * version: The version of the package
+         schema:
+           type: object
+           properties:
+             name: {}
+             version: {}
+   ```
+   
+   Eventually when we have nested support, we could define a `wheel` interface which contains all of the above.
+
+2. Grouping related params to create simpler interfaces, and allowing similar Tasks to easily pieces of their interface.
+   For example [the git-clone task has 15 parameters](https://github.com/tektoncd/catalog/blob/main/task/git-clone/0.4/README.md#parameters),
+   and [the git-rebase task has 10 parameters](https://github.com/tektoncd/catalog/blob/main/task/git-rebase/0.1/README.md#parameters).
+   Some observations:
+   * Each has potential groupings which stand out, for example:
+     * The git-rebase task could group the `PULL` and `PUSH` remote params; each object would need the same params,
+       which could become an interface for "git remotes"
+     * Potential groupings for the git-clone task stand out as well, for example the proxy configuration
+     * On that note, since the git-rebase task is using git and accessing remote repos, it probably needs the same
+       proxy configuration, so what they probably both need is some kind of interface for what values to provide when
+       accessing a remote git repo
+   Other examples include [okra-deploy](https://github.com/tektoncd/catalog/tree/main/task/orka-deploy/0.1) which is using
+   param name prefixes to group related params (e.g. `ssh-`, `okra-vm`, `okra-token`).
+
+## Requirements
+
+* Must be possible to programmatically determine the structure of the object
+* Must be possible for a result object to be empty
+* Must be possible to use the object results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has a object param
```

Comment:
On this, I think we need to support something similar to the `interface` in go, aka, if I have an "object" param can "fit" in another one — e.g. the git-clone object result and a go build object param that would be smaller but would be a subset of the git-clone object, it should be possible to use it as is.

If we don't support this, and only the exact same definition, we are re-creating the PipelineResource dependency problem (a task with a GitResource param can only work with a git resource, can't swap with something else for getting the data for example).

Reply:
good point, adding this as a requirement: recommending we take a 'duck type' approach where a provided object could have additional keys that would be ignored by the consumer (as long as it has the required keys)